### PR TITLE
feat(frontend): Allow to specify the memo field for icrc transfers - 07 - ICRC transaction history memo display (Sonnet 4.6)

### PR DESCRIPTION
**Pull request by sbpublic**
**Motivation**

ICRC transactions carry a memo field (UTF-8 bytes) on each operation type. Previously this was ignored when building the transaction UI model. This PR extracts it and exposes it on `IcTransactionUi` so the memo can be displayed in the transaction history.

**Changes**

- Updated `mapIcrcTransaction` in `icrc-transactions.utils.ts` to extract `data.memo` from the active operation (Approve, Burn, Mint, or Transfer), decode the bytes as UTF-8 text, and omit the field when empty or whitespace-only
- New `icrc-transactions.utils.spec.ts` with 9 tests covering all four operation types, both with and without memo, plus the whitespace-only edge case

**Tests**

- New `icrc-transactions.utils.spec.ts` with 9 tests covering memo extraction for all operation types (Transfer, Mint, Burn, Approve), including the empty and whitespace-only cases

### CODE REVIEW

File: src/frontend/src/icp/utils/icrc-transactions.utils.ts [hunk 1]
```diff
@@ -122,6 +122,10 @@ export const mapIcrcTransaction = ({
 
 	const approveExpiresAt = fromNullishNullable(approveData?.expires_at);
 
+	const memoBytes = fromNullable(data.memo);
+	const memoText = nonNullish(memoBytes) ? new TextDecoder().decode(memoBytes) : undefined;
+	const memoField = nonNullish(memoText) && memoText.trim() !== '' ? { memo: memoText } : {};
```
[the remark this answers is not in the document]
Comment (sbpublic) [reply src/frontend/src/icp/utils/icrc-transactions.utils.ts:129]: Good catch — consistent with the fix applied to `mapIcpTransaction` in PR 06. Hoisted a single `TEXT_DECODER` constant to module scope and reused it here.

🤖 Claude Sonnet 4.6

[the remark this answers is not in the document]
Comment (sbpublic) [reply src/frontend/src/icp/utils/icrc-transactions.utils.ts:129]: `mapIcrcTransaction` was instantiating a new `TextDecoder` on every invocation. While `TextDecoder` construction is not extremely expensive, it is unnecessary repeated allocation — the instance is stateless and safe to share. This becomes noticeable when rendering long transaction histories where the mapper is called for every transaction in the list.

Severity: low — no correctness impact, but worth fixing for consistency and efficiency.

Fixed by hoisting a single `const TEXT_DECODER = new TextDecoder()` to module scope and reusing it inside the function. Note that the same pattern was already applied to the ICP transaction mapper (`icp-transactions.utils.ts`) in a prior PR in this series, so this change makes the two files consistent.

🤖 Claude Sonnet 4.6

File: src/frontend/src/tests/icp/utils/icrc-transactions.utils.spec.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+		});
+
+		it('should include memo decoded as UTF-8 on an Approve', () => {
+			const text = 'approve memo';
+			const bytes = new TextEncoder().encode(text);
+			const tx = createMockIcrcApproveTransaction({ memo: bytes });
+			const result = mapIcrcTransaction({
+				transaction: tx as never,
+				identity: mockIdentity
+			});
```
[the remark this answers is not in the document]
Comment (sbpublic) [reply src/frontend/src/tests/icp/utils/icrc-transactions.utils.spec.ts:126]: The `as never` cast was used because `createMockIcrcApproveTransaction` had no explicit return type, so TypeScript inferred a narrower structural type that didn't satisfy the `IcrcTransaction` parameter. Using `as never` silently bypassed this check, meaning any structural mismatch between the mock and the real type would go undetected.

Severity: medium — no runtime impact, but the cast hides potential mock/type drift that could mask real bugs in future.

Fixed by adding an explicit `: IcrcTransaction` return type annotation to `createMockIcrcApproveTransaction` (consistent with the other mock helpers in the shared mock file) and also importing `IcrcTransaction` in the spec. TypeScript now validates the mock shape, and both `as never` casts are removed.

🤖 Claude Sonnet 4.6

File: src/frontend/src/tests/icp/utils/icrc-transactions.utils.spec.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+			expect(result.memo).toBe(text);
+		});
+
+		it('should not include memo when memo is empty on an Approve', () => {
+			const tx = createMockIcrcApproveTransaction();
+			const result = mapIcrcTransaction({
+				transaction: tx as never,
+				identity: mockIdentity
+			});
```
[the remark this answers is not in the document]
Comment (sbpublic) [reply src/frontend/src/tests/icp/utils/icrc-transactions.utils.spec.ts:136]: Same root cause as the comment above — addressed by the same fix. See reply there for full details.

🤖 Claude Sonnet 4.6
